src/proxy/http/timeout.rs

```diff
@@ -0,0 +1,111 @@
+use std::time::Duration;
+
+use futures::{future, Future, Poll};
+use http::{Request, Response, StatusCode};
+
+use svc::linkerd2_timeout::{Error as TimeoutError, Timeout};
+use svc;
+
+/// Implement on targets to determine if a service has a timeout.
+pub trait HasTimeout {
+    fn timeout(&self) -> Option<Duration>;
+}
+
+/// An HTTP-specific optional timeout layer.
+///
+/// The stack target must implement `HasTimeout`, and if a duration is
+/// specified for the target, a timeout is applied waiting for HTTP responses.
+///
+/// Timeout errors are translated into `http::Response`s with appropiate
+/// status codes.
+pub fn layer() -> Layer {
+    Layer
+}
+
+#[derive(Clone, Debug)]
+pub struct Layer;
+
+#[derive(Clone, Debug)]
+pub struct Stack<M> {
+    inner: M,
+}
+
+#[derive(Debug)]
+pub struct Service<S>(Timeout<S>);
+
+impl<T, M> svc::Layer<T, T, M> for Layer
+where
+    M: svc::Stack<T>,
+    T: HasTimeout,
+{
+    type Value = <Stack<M> as svc::Stack<T>>::Value;
+    type Error = <Stack<M> as svc::Stack<T>>::Error;
+    type Stack = Stack<M>;
+
+    fn bind(&self, inner: M) -> Self::Stack {
+        Stack {
+            inner,
+        }
+    }
+}
+
+impl<T, M> svc::Stack<T> for Stack<M>
+where
+    M: svc::Stack<T>,
+    T: HasTimeout,
+{
+    type Value = svc::Either<Service<M::Value>, M::Value>;
+    type Error = M::Error;
+
+    fn make(&self, target: &T) -> Result<Self::Value, Self::Error> {
+        let inner = self.inner.make(target)?;
+        if let Some(timeout) = target.timeout() {
+            Ok(svc::Either::A(Service(Timeout::new(inner, timeout))))
+        } else {
+            Ok(svc::Either::B(inner))
+        }
+    }
+}
+
+impl<S, B1, B2> svc::Service<Request<B1>> for Service<S>
+where
+    S: svc::Service<Request<B1>, Response = Response<B2>>,
+    B2: Default,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = future::OrElse<
+        <Timeout<S> as svc::Service<Request<B1>>>::Future,
+        Result<Response<B2>, S::Error>,
+        fn(TimeoutError<S::Error>) -> Result<Response<B2>, S::Error>,
+    >;
+
+    fn poll_ready(&mut self) -> Poll<(), Self::Error> {
+        match self.0.poll_ready() {
+            Ok(ok) => Ok(ok),
+            Err(TimeoutError::Error(err)) => Err(err),
+            Err(TimeoutError::Timeout(_)) |
+            Err(TimeoutError::Timer(_)) => unreachable!("timeout error in poll_ready"),
+        }
+    }
+
+    fn call(&mut self, req: Request<B1>) -> Self::Future {
+        self.0.call(req).or_else(|err| match err {
+            TimeoutError::Error(err) => Err(err),
+            TimeoutError::Timeout(dur) => {
+                debug!("request timed out after {:?}", dur);
+                let mut res = Response::default();
+                *res.status_mut() = StatusCode::GATEWAY_TIMEOUT;
```

Comment:
Things like finagle return a 503, but usually finagle is used as *part of* the actual service. Since we're a proxy, it might be more appropriate to return 504, which is described as a timeout encountered inside a proxy. I could be convinced either way.